apiserver/facades/controller/instancepoller/merge.go

```diff
@@ -133,34 +137,27 @@ func (o *mergeMachineLinkLayerOp) opsForDeviceOriginRelinquishment(
 }
 
 func (o *mergeMachineLinkLayerOp) processExistingDeviceAddress(
-	addr networkingcommon.LinkLayerAddress, incomingDev network.InterfaceInfo,
+	addr networkingcommon.LinkLayerAddress, incomingAddrs []state.LinkLayerDeviceAddress,
 ) ([]txn.Op, error) {
 	addrValue := addr.Value()
 
-	// TODO (manadart 2020-06-09): This is where we see a cardinality mismatch.
-	// The InterfaceInfo type has one provider address ID, but a collection of
-	// addresses and shadow addresses.
-	// This should change so that addresses are collections of types that
-	// include a provider ID, instead of assuming that the provider ID applies
-	// to the primary address (index 0).
-	// We should also be adding shadow addresses to the link-layer device here.
-	// For now, if the provider does not recognise the address,
-	// give responsibility for it to the machine.
-	if !set.NewStrings(incomingDev.Addresses.ToIPAddresses()...).Contains(addrValue) {
-		return addr.SetOriginOps(network.OriginMachine), nil
-	}
-
-	// If the address is the incoming primary address, assign the provider IDs.
-	// Otherwise simply indicate that the provider is aware of this address.
-	if addrValue == incomingDev.PrimaryAddress().Value {
-		ops, err := addr.SetProviderIDOps(incomingDev.ProviderAddressId)
-		if err != nil {
-			return nil, errors.Trace(err)
+	// If one of the incoming addresses matches the existing one,
+	// return ops for setting the incoming provider IDs.
+	for _, incomingAddr := range incomingAddrs {
+		if strings.HasPrefix(incomingAddr.CIDRAddress, addrValue) {
```

Comment:
Don't you need a `&& !o.processedAddrs.Contains(addrValue)` here? Also you could invert the if condition and continue so you can remove one indentation level from the following block.